### PR TITLE
Handle BinaryExpressions in pure mode

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

This greatly improves the component folding in the Hacker News test, where before it would fail due to a FatalError in BinaryExpressions. In a pure wrapper we can safely proceed with this.

### CODE REVIEW

**Comment by gaearon** [comment]
>In a pure wrapper we can safely proceed with this.

Can you explain your reasoning behind this?

I understand both things are called "pure" but I'm not sure they refer to the same concept and are necessarily safe.

The definition of purity in this file:

https://github.com/trueadm/prepack/blob/e007604a6358b9195c7ea93ef3648916d068921d/src/evaluators/BinaryExpression.js#L51

>(terminates, does not throw exception, does not read or write heap)

Can you explain why neither of those cases is relevant in pure scope?

File: src/evaluators/BinaryExpression.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ export function getPureBinaryOperationResultType(
   function reportErrorIfNotPure(purityTest: (Realm, Value) => boolean, typeIfPure: typeof Value): typeof Value {
     let leftPure = purityTest(realm, lval);
     let rightPure = purityTest(realm, rval);
-    if (leftPure && rightPure) return typeIfPure;
+    if ((leftPure && rightPure) || realm.isInPureScope()) return typeIfPure;
```
Comment (gaearon) [inline src/evaluators/BinaryExpression.js]: I think this makes sense but maybe it needs to be:

```js
if (leftPure && rightPure) return typeIfPure;

if (realm.isInPureScope()) {
  if (!leftPure) Leak.leakValue(lval);
  if (!rightPure) Leak.leakValue(rval);
  return typeIfPure;
}
```

Similar to this: https://github.com/facebook/prepack/blob/4f22fbae831e08804104f1b08bcd8b5ad00c44ee/src/values/AbstractObjectValue.js#L346

?

Because they might leak the object to `toString()` and `valueOf()`.

I'm struggling to come up with a case where that would matter though.

Comment (sebmarkbage) [reply src/evaluators/BinaryExpression.js]: Yes, and also evaluateWithPossibleThrowCompletion since it might throw.

**Comment by trueadm** [comment]
@gaearon I addressed the PR feedback and hopefully what I changed is what you meant by your points above.

**Comment by gaearon** [comment]
Sebastian suggested we should be more specific than that. It’s not necessary to always leak everything.

Instead we should look at specific operators (+, -, typeof, etc) and see in which cases they would call into user code. Such as + calling valueOf. Then leak the specific thing we need to leak just before it happens.

**Comment by trueadm** [comment]
@gaearon I didn't quite understand that point as `reportErrorIfNotPure` is only called via certain paths, the ones I saw were:

`<`, `>`, `>=`, `<=`, `>>>`, `<<`, `>>`, `&`, `|`, `^`, `**`, `%`, `/`, `*`, `-`

The other operators don't appear to hit that code path, unless I'm missing something.

File: src/evaluators/BinaryExpression.js [hunk 2]
```diff
@@ -61,6 +61,13 @@ export function getPureBinaryOperationResultType(
     let leftPure = purityTest(realm, lval);
     let rightPure = purityTest(realm, rval);
     if (leftPure && rightPure) return typeIfPure;
+
+    if (realm.isInPureScope()) {
+      if (!leftPure) Leak.leakValue(realm, lval);
```
Comment (NTillmann) [inline src/evaluators/BinaryExpression.js]: Consider renaming the enclosing function to `leakOrReportErrorIfNotPure

Comment (trueadm) [reply src/evaluators/BinaryExpression.js]: Done, are you happy with this PR apart from that?

**Comment by gaearon** [comment]
This PR doesn't contain a serializer test. I would prefer if future changes like this did because if there is some edge case we need to handle, it's not obvious which React test I should be looking at. It is also likely that we'll replace React tests more often and lose coverage for this.

I tried to start by adding a very simple serializer test:

```js
// additional functions
// abstract effects

let obj1 = global.__abstract ? __abstract('object', '({foo: {valueOf() { return 42; }}})') : {foo: {valueOf() { return 42; }}};
let obj2 = global.__abstract ? __abstract('object', '({foo: {bar: {valueOf() { return 42; }}}})') : {foo: {bar: {valueOf() { return 42; }}}};
if (global.__makeSimple) {
  __makeSimple(obj2);
}

function additional1() {
  return '' + obj1.foo;
}

function additional2() {
  return '' + obj2.foo.bar;
}

inspect = function() {
  let ret1 = additional1();
  let ret2 = additional2();
  return ret1 + ret2;
}
```

Unfortunately it gives me a Node `SIGSEGV` on this branch. The master just fatals (as expected).

Does this happen to anybody else?

**Comment by trueadm** [comment]
@gaearon Sorry, we can add any issues in follow up PRs.
